### ISSUE TITLE
For better AI "example" but it work, full version Contac me

### ISSUE THREAD

**Issue by antorizwar0201-wq** [antorizwar0201-wq/Suprianto#1]
# HSS AdaptiveLR Lite v1.0
# By Suprianto - HSS Lab Indonesia
# License: MIT

import torch
from torch.optim.lr_scheduler import _LRScheduler

class AdaptiveLRLite(_LRScheduler):
    def __init__(self, optimizer, patience=5, factor=0.5, min_lr=1e-6):
        self.patience = patience
        self.factor = factor
        self.min_lr = min_lr
        self.counter = 0
        self.best = None
        super().__init__(optimizer)

    def step(self, metrics):
        # metrics = validation loss
        if self.best is None or metrics < self.best:
            self.best = metrics
            self.counter = 0
        else:
            self.counter += 1

        if self.counter >= self.patience:
            self._reduce_lr()
            self.counter = 0

    def _reduce_lr(self):
        for i, param_group in enumerate(self.optimizer.param_groups):
            old_lr = float(param_group['lr'])
            new_lr = max(old_lr * self.factor, self.min_lr)
            param_group['lr'] = new_lr

**Comment by antorizwar0201-wq** [comment]
"""
HSS-AdaptiveLR-Lite v1.0
A Self-Aware Learning Rate Scheduler for PyTorch
Created by: Suprianto - HSS Lab, Indonesia
License: MIT
GitHub: https://github.com/HSS-Lab/AdaptiveLR-Lite

Description:
Unlike fixed schedulers, this scheduler "watches" your model's validation loss.
If the model stops improving for 'patience' epochs, it automatically reduces the LR.
Result: More stable training and anti-overfitting.
"""

import torch
from torch.optim.lr_scheduler import _LRScheduler

class AdaptiveLRLite(_LRScheduler):
    def __init__(self, optimizer, patience=5, factor=0.5, min_lr=1e-6, verbose=False):
        """
        Args:
            optimizer (Optimizer): Wrapped optimizer.
            patience (int): Number of epochs with no improvement after which lr will be reduced. Default: 5
            factor (float): Factor by which the learning rate will be reduced. new_lr = lr * factor. Default: 0.5
            min_lr (float): A lower bound on the learning rate of all param groups. Default: 1e-6
            verbose (bool): If True, prints a message to stdout for each reduction. Default: False
        """
        self.patience = patience
        self.factor = factor
        self.min_lr = min_lr
        self.verbose = verbose
        
        self.counter = 0
        self.best = None
        self.num_bad_epochs = 0

        super(

**Comment by antorizwar0201-wq** [comment]
Python kode

from hss_adaptive_lr import AdaptiveLRLite

optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
scheduler = AdaptiveLRLite(optimizer, patience=5, factor=0.5, verbose=True)

for epoch in range(100):
    train(...)
    val_loss = validate(...)
    
    scheduler.step(val_loss) # <- Cukup panggil ini tiap epoch
